Userland/Libraries/LibJS/Parser.cpp

```diff
@@ -2173,16 +2227,27 @@ NonnullRefPtr<FunctionNodeType> Parser::parse_function_node(u8 parse_options)
     TemporaryChange might_need_arguments_object_rollback(m_state.function_might_need_arguments_object, false);
 
     constexpr auto is_function_expression = IsSame<FunctionNodeType, FunctionExpression>;
-    auto function_kind = (parse_options & FunctionNodeParseOptions::IsGeneratorFunction) != 0 ? FunctionKind::Generator : FunctionKind::Regular;
+    FunctionKind function_kind;
+    if ((parse_options & FunctionNodeParseOptions::IsGeneratorFunction) != 0 && (parse_options & FunctionNodeParseOptions::IsAsyncFunction) != 0)
+        TODO();
+    else if ((parse_options & FunctionNodeParseOptions::IsGeneratorFunction) != 0)
+        function_kind = FunctionKind::Generator;
+    else if ((parse_options & FunctionNodeParseOptions::IsAsyncFunction) != 0)
+        function_kind = FunctionKind::Async;
+    else
+        function_kind = FunctionKind::Regular;
     String name;
     if (parse_options & FunctionNodeParseOptions::CheckForFunctionAndName) {
+        if (function_kind == FunctionKind::Regular && match(TokenType::Async)) {
```

Comment:
We have to check that the token after `async` does not contain a line terminator as then ASI kicks in and makes the function non-async.
```js
async // this line ending is important
function f() {
  await 3; // <- should fail since f is not async
}
```

Reply:
great catch, fixed